### PR TITLE
Allow evm to call native modules through logs

### LINKED ISSUE #416: Allow evm to call native modules through logs
__Proposal:__ Allow evm to call native modules through logs

https://github.com/tharsis/ethermint/discussions/394

__Current behavior:__ evm can't call native modules

__Desired behavior:__ allow evm to call native modules

__Use case:__ 
- manage native tokens
  - ibc tokens
  - gravity bridge
- many potential

### PR DESCRIPTION
**Pull request by yihuang**
Closes #416

## Description

- Add `EvmHooks`, currently only single api `PostTxProcessing`.
- implements a `BankSendHook` which can send native coins, but not enabled in app by default.

```solidity
    function test() public {
        emit __CosmosNativeBankSend(msg.sender, 10, "aphoton");
    }
```

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by fedekunze** [comment]
@yihuang let's create an ADR for this that leverages Hooks. See the governance hooks for reference

File: x/evm/types/expected_keepers.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ethcmn "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+// Event Hooks
+// These can be utilized to customize evm transaction processing.
+
+// EvmHooks event hooks for governance proposal object (noalias)
+type EvmHooks interface {
```
Comment (fedekunze) [inline x/evm/types/expected_keepers.go:13]: move to `interfaces.go`

File: x/evm/types/hooks.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+var (
+	// BankSendEvent represent the signature of
+	// `event __CosmosNativeBankSend(address recipient, uint256 amount, string denom)`
+	BankSendEvent abi.Event
+
+	_ EvmHooks = MultiEvmHooks{}
+	_ EvmHooks = BankSendHook{}
+)
+
+// MultiEvmHooks combine multiple governance hooks, all hook functions are run in array sequence
```
Comment (fedekunze) [inline x/evm/types/hooks.go:22]: replace `governance` with `EVM`

File: x/evm/types/hooks.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+type MultiEvmHooks []EvmHooks
```
Comment (fedekunze) [inline x/evm/types/hooks.go:23]: Imo we should also register the module name on MultiEVMHooks for better error messages

File: x/evm/types/hooks.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+// NewMultiEvmHooks combine multiple governance hooks
+func NewMultiEvmHooks(hooks ...EvmHooks) MultiEvmHooks {
+	return hooks
+}
+
+// PostTxProcessing delegate the call to underlying hooks
+func (h MultiEvmHooks) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
```
Comment (fedekunze) [inline x/evm/types/hooks.go:31]: nit
```suggestion
func (mh MultiEvmHooks) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
```

File: x/evm/types/hooks.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	for i := range h {
+		err := h[i].PostTxProcessing(ctx, txHash, logs)
+		if err != nil {
+			return err
```
Comment (fedekunze) [inline x/evm/types/hooks.go:35]:
```suggestion
	for i := range mh {
		if err := mh[i].hook.PostTxProcessing(ctx, txHash, logs); err != nil {
			return sdkerrors.Wrapf(err, "module %s EVM hooks failed", mh[i].module)
```

File: x/evm/types/hooks.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		}, abi.Argument{
+			Name:    "denom",
+			Type:    stringType,
+			Indexed: false,
+		}},
+	)
+}
+
+// BankSendHook handle the `__CosmosNativeBankSend` log.
+type BankSendHook struct {
```
Comment (fedekunze) [inline x/evm/types/hooks.go:66]: the BankSendHook shouldn't be included in this PR or module

File: x/evm/types/hooks.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+// NewBankSendHook creates a new bank send hook
+func NewBankSendHook(bankKeeper bankkeeper.Keeper) *BankSendHook {
+	return &BankSendHook{
+		bankKeeper: bankKeeper,
+	}
+}
+
+// PostTxProcessing implements EvmHooks interface
+func (h BankSendHook) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
```
Comment (fedekunze) [inline x/evm/types/hooks.go:78]: remove and include in another PR

File: x/evm/keeper/keeper_test.go [hunk 8]
```diff
@@ -228,3 +201,45 @@ func (suite *KeeperTestSuite) DeployTestContract(t require.TestingT, owner commo
 func TestKeeperTestSuite(t *testing.T) {
 	suite.Run(t, new(KeeperTestSuite))
 }
+
+func (suite *KeeperTestSuite) TestBankHooks() {
```
Comment (fedekunze) [inline x/evm/keeper/keeper_test.go:205]: this test could be a dummy EVM hook and for the follow up PR we can add this on the new module that implements this hook

File: x/evm/keeper/state_transition.go [hunk 9]
```diff
@@ -181,6 +184,16 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 
 	res.Hash = txHash.Hex()
 	logs := k.GetTxLogs(txHash)
+
+	if !res.Failed() {
+		err = k.PostTxProcessing(txHash, logs)
+		if err != nil {
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:190]: also add the comments from the ADR
```suggestion
		if err := k.PostTxProcessing(txHash, logs); err != nil {
```

File: x/evm/keeper/state_transition.go [hunk 10]
```diff
@@ -181,6 +184,16 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
...
+			k.RevertToSnapshot(revision)
+			res.VmError = "failed to process native logs"
+			res.Ret = []byte(err.Error())
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:193]: is this how the error is returned on geth (i.e just the byte representation of the error message)?

File: x/evm/keeper/state_transition.go [hunk 11]
```diff
@@ -181,6 +184,16 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 
 	res.Hash = txHash.Hex()
 	logs := k.GetTxLogs(txHash)
+
+	if !res.Failed() {
+		err = k.PostTxProcessing(txHash, logs)
+		if err != nil {
+			k.RevertToSnapshot(revision)
+			res.VmError = "failed to process native logs"
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:192]: let's make this a constant and move it to `errors.go` so that we can check for this error message on tests

**Review by fedekunze — CHANGES_REQUESTED** [review]

File: x/evm/types/hooks.go (hunk 3 above)
Comment (fedekunze) [reply x/evm/types/hooks.go:23]: also let's move this file to the `keeper/` directory

File: x/evm/types/hooks.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/hooks.go:35]: rather than add a new method to the interface, how about just report the reflected type name:
```
return sdkerrors.Wrapf(err, "EVM hook %s failed", reflect.TypeOf(mh[i]))
```

File: x/evm/keeper/state_transition.go (hunk 10 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go:193]: Only `ErrExecutionReverted` put error reason in the `Ret` field, it's encoded in ABI format, other VM errors don't have error reasons.

File: x/evm/keeper/state_transition.go (hunk 11 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go]: done

File: x/evm/keeper/state_transition.go (hunk 10 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go]: done

File: x/evm/keeper/state_transition.go (hunk 9 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go]: done

File: x/evm/keeper/keeper_test.go (hunk 8 above)
Comment (yihuang) [reply x/evm/keeper/keeper_test.go]: done

File: x/evm/types/hooks.go (hunk 7 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/hooks.go (hunk 6 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/hooks.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/hooks.go (hunk 4 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/hooks.go (hunk 3 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/hooks.go (hunk 2 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

File: x/evm/types/expected_keepers.go (hunk 1 above)
Comment (yihuang) [reply x/evm/types/expected_keepers.go]: done

File: x/evm/types/hooks.go (hunk 5 above)
Comment (fedekunze) [reply x/evm/types/hooks.go]: let's use `%T` instead of `reflect`

File: x/evm/keeper/hooks.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+// NewMultiEvmHooks combine multiple evm hooks
+func NewMultiEvmHooks(hooks ...types.EvmHooks) MultiEvmHooks {
+	return hooks
+}
+
+// PostTxProcessing delegate the call to underlying hooks
+func (mh MultiEvmHooks) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
+	for i := range mh {
+		if err := mh[i].PostTxProcessing(ctx, txHash, logs); err != nil {
+			return sdkerrors.Wrapf(err, "EVM hook %s failed", reflect.TypeOf(mh[i]))
```
Comment (fedekunze) [inline x/evm/keeper/hooks.go:29]:
```suggestion
			return sdkerrors.Wrapf(err, "EVM hook %T failed", mh[i])
```

File: x/evm/keeper/hooks_test.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	return nil
+}
+
+// FailureHook always fail
+type FailureHook struct{}
+
+func (dh FailureHook) PostTxProcessing(ctx sdk.Context, txHash common.Hash, logs []*ethtypes.Log) error {
+	return errors.New("post tx processing failed")
+}
+func (suite *KeeperTestSuite) TestEvmHooks() {
```
Comment (fedekunze) [inline x/evm/keeper/hooks_test.go:32]:
```suggestion
}

func (suite *KeeperTestSuite) TestEvmHooks() {
```

File: x/evm/keeper/hooks_test.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	logs := k.GetTxLogs(txHash)
+	suite.Require().Equal(1, len(logs))
+
+	err = k.PostTxProcessing(txHash, logs)
+	suite.Require().NoError(err)
+
+	suite.Require().Equal(1, len(logRecordHook.Logs))
+}
+
+func (suite *KeeperTestSuite) TestHookFailure() {
```
Comment (fedekunze) [inline x/evm/keeper/hooks_test.go:64]: can we use Table Tests instead of the 2 test functions? See this [code](https://github.com/tharsis/ethermint/pull/515/files#diff-86a2788c6f900e3e8d070486fa1b357f309c85c581bb1b7b64597b7d7445cc8dR19) for reference

File: x/evm/keeper/keeper.go [hunk 15]
```diff
@@ -416,3 +419,21 @@ func (k Keeper) ResetAccount(addr common.Address) {
...
+func (k *Keeper) SetHooks(eh types.EvmHooks) *Keeper {
+	if k.hooks != nil {
+		panic("cannot set evm hooks twice")
+	}
+
+	k.hooks = eh
+	return k
+}
+
+// PostTxProcessing delegate the call to the hooks
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:433]:
```suggestion
// PostTxProcessing delegate the call to the hooks. If no hook has been registered, this function returns with a `nil` error
```

File: x/evm/keeper/keeper.go [hunk 16]
```diff
@@ -416,3 +419,21 @@ func (k Keeper) ResetAccount(addr common.Address) {
 	k.DeleteCode(addr)
 	k.DeleteAccountStorage(addr)
 }
+
+// SetHooks sets the hooks for governance
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:423]:
```suggestion
// SetHooks sets the hooks for the EVM module
```

File: x/evm/keeper/keeper.go [hunk 17]
```diff
@@ -54,6 +54,9 @@ type Keeper struct {
 	// trace EVM state transition execution. This value is obtained from the `--trace` flag.
 	// For more info check https://geth.ethereum.org/docs/dapp/tracing
 	debug bool
+
+	// EvmHooks
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:58]:
```suggestion
	// EVM Hooks for tx post-processing
```

File: x/evm/keeper/state_transition.go [hunk 18]
```diff
@@ -175,6 +175,9 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 		panic("context stack shouldn't be dirty before apply message")
 	}
 
+	// Contains the tx processing and post processing in same scope
+	revision := k.Snapshot()
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:179]: I thought this was done internally at during the EVM call/create, or if we do that it doesn't include the post-processing?

Can you update the ADR and add this line?

File: x/evm/types/interfaces.go [hunk 19]
```diff
@@ -33,3 +35,12 @@ type StakingKeeper interface {
 	GetHistoricalInfo(ctx sdk.Context, height int64) (stakingtypes.HistoricalInfo, bool)
 	GetValidatorByConsAddr(ctx sdk.Context, consAddr sdk.ConsAddress) (validator stakingtypes.Validator, found bool)
 }
+
+// Event Hooks
+// These can be utilized to customize evm transaction processing.
+
+// EvmHooks event hooks for evm tx processing
+type EvmHooks interface {
+	// Must be called after tx is processed, if failed, the whole evm transaction is reverted.
```
Comment (fedekunze) [inline x/evm/types/interfaces.go:44]: Add the warning that we discussed in the `Negative` consequences section of the ADR

**Review by fedekunze — COMMENTED** [review]
Final pass. Great work! Can you also add a new `Features` entry on the Changelog? 🙏

File: x/evm/keeper/state_transition.go (hunk 18 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go:179]: > I thought this was done internally at during the EVM call/create, or if we do that it doesn't include the post-processing?

Yes, we need this to include the post-processing step.

> 
> Can you update the ADR and add this line?

sure

File: x/evm/types/interfaces.go (hunk 19 above)
Comment (yihuang) [reply x/evm/types/interfaces.go:44]: Actually, that negative consequence is about the use case of utilizing the hook to process logs, not about the hook interface in general.

File: x/evm/keeper/state_transition.go (hunk 18 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go:179]: done

File: x/evm/keeper/hooks_test.go (hunk 14 above)
Comment (yihuang) [reply x/evm/keeper/hooks_test.go]: done

File: x/evm/types/interfaces.go (hunk 19 above)
Comment (yihuang) [reply x/evm/types/interfaces.go]: I added the use case context to consequences in the adr.

File: x/evm/types/hooks.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/hooks.go]: done

**Comment by yihuang** [comment]
> Final pass. Great work! Can you also add a new `Features` entry on the Changelog? 🙏

changelog added.
